source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,193 @@
+    if (end_stream) {
+      return;
+    }
+    if (result.content_length_ > 0) {
+      remaining_body_.emplace_back(0, result.content_length_);
+      getBody();
+    } else {
+      lookup_->getTrailers(
+          [this](Http::ResponseTrailerMapPtr&& trailers) { onTrailersAsync(std::move(trailers)); });
+    }
+  }
+}
+
+void CacheFilter::onHeadersAsync(LookupResult&& result) {
+  post([this, status = result.cache_entry_status_, headers = result.headers_.release(),
+        response_ranges = std::move(result.response_ranges_),
+        content_length = result.content_length_, has_trailers = result.has_trailers_] {
+    onHeaders(LookupResult{status, absl::WrapUnique(headers), content_length, response_ranges,
+                           has_trailers});
+  });
+}
+
+void CacheFilter::getBody() {
+  ASSERT(!remaining_body_.empty(), "No reason to call getBody when there's no body to get.");
+  lookup_->getBody(remaining_body_[0],
+                   [this](Buffer::InstancePtr&& body) { onBody(std::move(body)); });
+}
+
+void CacheFilter::onBodyAsync(Buffer::InstancePtr&& body) {
+  post([this, body = body.release()] { onBody(absl::WrapUnique(body)); });
+}
+
+// TODO(toddmgreer): Handle downstream backpressure.
+void CacheFilter::onBody(Buffer::InstancePtr&& body) {
+  if (!active()) {
+    return;
+  }
+  if (remaining_body_.empty()) {
+    ASSERT(false, "CacheFilter doesn't call getBody unless there's more body to get, so this is a "
+                  "bogus callback.");
+    decoder_callbacks_->resetStream();
+    return;
+  }
+
+  if (!body) {
+    ASSERT(false, "Cache said it had a body, but isn't giving it to us.");
+    decoder_callbacks_->resetStream();
+    return;
+  }
+
+  const uint64_t bytes_from_cache = body->length();
+  if (bytes_from_cache < remaining_body_[0].length()) {
+    remaining_body_[0].trimFront(bytes_from_cache);
+  } else if (bytes_from_cache == remaining_body_[0].length()) {
+    remaining_body_.erase(remaining_body_.begin());
+  } else {
+    ASSERT(false, "Received oversized body from cache.");
+    decoder_callbacks_->resetStream();
+    return;
+  }
+
+  const bool end_stream = remaining_body_.empty() && !response_has_trailers_;
+  decoder_callbacks_->encodeData(*body, end_stream);
+  if (!remaining_body_.empty()) {
+    getBody();
+  } else if (response_has_trailers_) {
+    lookup_->getTrailers(
+        [this](Http::ResponseTrailerMapPtr&& trailers) { onTrailersAsync(std::move(trailers)); });
+  }
+}
+
+void CacheFilter::onTrailers(Http::ResponseTrailerMapPtr&& trailers) {
+  if (active()) {
+    decoder_callbacks_->encodeTrailers(std::move(trailers));
+  }
+}
+
+void CacheFilter::onTrailersAsync(Http::ResponseTrailerMapPtr&& trailers) {
+  post([this, trailers = trailers.release()] { onTrailers(absl::WrapUnique(trailers)); });
+}
+
+void CacheFilter::post(std::function<void()> f) const {
+  decoder_callbacks_->dispatcher().post(std::move(f));
+}
```

Comment:
Sorry why are we doing the post pattern here? Is there an assumption that the callbacks can be run on a different thread? Would it be better/easier to assume that the callbacks always run on the same thread as the requester much like the rest of Envoy? This would also make things a bit simpler around destruction and cleanup per my comment above. I think I know see why you need to keep track of the decoder callbacks. But even so, I don't think this is safe. Post callbacks can't be cancelled and you can race between the post and destruction.

Reply:
Yes, callbacks might run on different threads. HttpCache implementations that store content in external services often have to handle the completion of those operations in arbitrary threads--one pattern is to multiplex all requests for a particular instance through a single connection managed on a single thread. Either CacheFilter needs to handle getting back to the right worker thread, or each asynchronous HttpCache implementation must do it. To me, it seems better to handle that complexity in one place (CacheFilter), rather than in several different implementations, each of which might introduce different bugs.

Reply:
My advice here would be to assume the general Envoy programming model of per-thread operations here like we do in almost all other places, otherwise this is going to get a lot more complicated than it needs to be right now. I would recommend revisiting later as needed. If you go with this solution, the current code isn't actually safe and we would need to either go back to using shared_from_this and/or implement cancellable posts through the dispatcher (preferred). Either way I don't think it's worth the extra complexity right now.

Reply:
The use of decoder_callbacks_ in post is a race (leftover from the prior form of the code, where it was ok), which I'll fix by clearing it after the calls to (Lookup|Insert)Context::onDestroy.

All callbacks, and thus all calls to post, are required to happen before onDestroy completes. IIUC, deletion is then deferred until after currently queued actions on the current dispatcher have finished, which implies that all posts finish before deletion.

That should be sufficient to avoid races, unless I'm missing something.

Reply:
You can race with the post calls, which is why I suggest you remove all the posting for now. I don't think it's necessary.

Reply:
Removed support for off-thread HttpCache implementations--it is now their responsibility to execute callbacks on the right thread.

Reply:
I had to add some state to CacheFilter to properly handle callbacks both during and after the call to decodeHeaders.